include/picongpu/simulation/stage/SynchrotronRadiation.hpp

```diff
@@ -0,0 +1,279 @@
+                }
+            }
+            return zq * integral;
+        }
+
+        //! @note see paper: "Extended particle-in-cell schemes for physics in ultrastrong laser fields: Review and
+        //! developments" by A. Gonoskov et.Al.
+        static float_64 secondSynchrotronFunction(float_64 const x)
+        {
+            return x * std::cyl_bessel_k(2. / 3, x);
+        }
+
+    public:
+        /** Create a synchrotron radiation instance: precompute first and second synchrotron functions
+         *
+         * @param cellDescription mapping for kernels
+         */
+        SynchrotronRadiation(MappingDesc const cellDescription) : cellDescription(cellDescription)
+        {
+            using namespace picongpu::particles::synchrotron::params; // for "InterpolationParams" struct
+            auto data_space = DataSpace<2>{InterpolationParams::numberTableEntries, 2};
+            auto grid_layout = GridLayout<2>{data_space};
+
+            // capture the space for table and variables
+            tableValuesF1F2 = std::make_shared<GridBuffer<float_X, 2>>(grid_layout);
+            failedRequirementQ = std::make_shared<GridBuffer<int32_t, 1>>(DataSpace<1>{1});
+            // set the values of variables to false
+            failedRequirementQ->getHostBuffer().getDataBox()(DataSpace<1>{0}) = false;
+            failedRequirementPrinted = false;
+
+            constexpr float_64 minZqExp = InterpolationParams::minZqExponent;
+            constexpr float_64 maxZqExp = InterpolationParams::maxZqExponent;
+            constexpr float_64 tableEntries = InterpolationParams::numberTableEntries;
+
+            // precompute F1 and F2 on log scale
+            for(uint32_t iZq = 0; iZq < tableEntries; iZq++)
+            {
+                float_64 zq = std::pow(2, minZqExp + (maxZqExp - minZqExp) * iZq / (tableEntries - 1));
+                // inverse function for index retrieval:
+                // index = (log2(zq) - minZqExp) / (maxZqExp - minZqExp) * (tableEntries-1);
+
+                float_64 const F1 = firstSynchrotronFunction(zq);
+                float_64 const F2 = secondSynchrotronFunction(zq);
+
+                tableValuesF1F2->getHostBuffer().getDataBox()(DataSpace<2>{iZq, u32(Accessor::f1)})
+                    = static_cast<float_X>(F1);
+                tableValuesF1F2->getHostBuffer().getDataBox()(DataSpace<2>{iZq, u32(Accessor::f2)})
+                    = static_cast<float_X>(F2);
+            }
+            // move the data to the device
+            tableValuesF1F2->hostToDevice();
+            failedRequirementQ->hostToDevice();
+
+            //! debug only, print F1 and F2, @todo remove
+            if constexpr(picongpu::particles::synchrotron::T_Debug)
+            {
+                int timesMore = 1;
+                for(uint32_t zq = 0; zq < InterpolationParams::numberTableEntries * timesMore; zq++)
+                {
+                    float_64 zq_ = std::pow(
+                        2,
```

Comment:
should this be integer or double precision? If it should be double please use `2.0`

Reply:
double
done